lib/charms/karma_k8s/v0/karma.py

```diff
@@ -86,7 +86,7 @@ class KarmaProvider(ProviderBase):
 
     ```python
     self.provider = KarmaProvider(
-        self, "karma-dashboard", "karma", "0.0.1"
+        self, "karma-dashboard", "karma", "0.86"
```

Comment:
May want to make it `>=` or `<=` just for safety

Reply:
the ProviderBase doc string implies it should be a fixed number:

```
        version: a string providing version of service provided by the
            charm. [...]
            It is important that the version is obtained by actually
            querrying the deployed application rather than being written
            into the code. This is because the same charm may be used
            to deploy different versions of a service (application).
```

Reply:
It does support comparators, though.

Reply:
semver does, but from OF POV, the version the provider reports needs to be the version of the service/executable, i.e. the output of the `karma --version`. tbh I do not understand why would it be sensible to construct a provider with non-fixed version.

Reply:
OF does not care. This is used for checking only whether the provider and consumer are compatible. Updating this string for every revision and finding that a consumer which specifies `0.87` is not actually compatible even though the API has not changed at all is going to be an annoying, frustrating experience. The version matters for this reason. The goal of `LIBAPI` is to say “hey, the library API is different”. The goal of the version in `ops.relation` is to say “ok, the library API is the same, but the workload container changed its configuration format, and the data we give you back will cause it to fail to start”. 

The exact version is not important. The operator framework does not (and arguably cannot) verify this, for the same reason as it does not support a “hot reload”. The flags to check it from the binary may be different, not exposed at all, etc. In addition, this is not exposed in `juju status`, so it has no meaning to users or administrators. 

https://github.com/canonical/operator/blob/92c7922e451b14184ceb9e7170d1564306fa1046/ops/relation.py#L482

I’d ask myself “do I want to submit a new PR here and force external consumers of the library to update theirs every single time there is a version bump, even if it doesn’t chance the `Provider|Consumer` API and the workload API|configuration syntax hasn’t changed at all”?

Reply:
I don't see where the problem is: provider says "I am 0.86" (currently hard-coded, but in the future will be taken from container app) and consumer needs ">=0.86" so even if the provider workload is newer, it will be accepted by the consumer because ">=0.86".

The code you linked seems to confirm this:
```python
        spec = semver.SimpleSpec(self._normalized_version(needs))
        got = semver.Version.coerce(self._normalized_version(has))

        return spec.match(got)
```

My understanding of the provider version may be different (please correct me where I am wrong):
1. When the Provider is instantiated, the exact (and correct) version of the workload app needs to be passed to it.
2. Internally, the Provider lib can behave differently, depending on the version of the app.
3. Internally, the Consumer lib can behave differently, depending on the version of the app.
4. If the Consumer version string is conservative (i.e. not open ended such as ">=x.y.z") then the version should indeed be bumped.

Reply:
1. "needs" is very prescriptive. Again, the provider does not (and cannot) check the exact version of the workload app, since the OF does not contain a mechanism to know how to check this (API call? CLI flag? Buildinfo stored elsewhere? It's workload dependent). The version is _relatively_ arbitrary and can be used by you, the Charm author, to decide whether the provider and consumer are compatible. The guidance in the OF docs is not telling you "the operator framework will check this", it's telling you "you should [check this yourself](https://github.com/canonical/prometheus-operator/blob/c57508523dda7e6019ca718a133e6542b51fa44f/src/charm.py#L406) and [use it in the constructor](https://github.com/canonical/prometheus-operator/blob/c57508523dda7e6019ca718a133e6542b51fa44f/src/charm.py#L54) because someone may specify a different resource which makes your in-code constraints meaningless", not that the OF will check it.

2. It can behave differently, but should not. In practice, the reason for different `LIBAPI`, `LIBPATCH`, and `version` (in ProviderBase/ConsumerBase) is to specify a level of compatibility, where we can enforce that the provider/consumer have matching APIs and semantics, similar to `SemVer` in Golang and other implementations. If the Provider would need different behavior, constraints should be added so the consumer matches the expected behavior rather than having conditional behavior which depends on the version.

3. Consumer is essentially the same as 2.

4. Yes. It's worth pointing out, though, that the goal of using `SemVer` is to follow the Semantic Versioning paradigm. That is: the goal is to provide you, the library author, with flexibility in versioning to avoid "dependency hell" by providing some flexibility. The versioning scheme is completely up to you, as long as it is consistent. For whatever warts the system does have, the [go module versioning](https://blog.golang.org/publishing-go-modules) guidelines provide a good rationale under "semantic versioning and modules".

As "best practice", the charm code (not OF) should query the running application and use that in the constructor, not hardcoding it. My comment with comparators is saying "if you are going to hardcode it because AlertManager is prerelease (v0.x) software, the provider could be more flexible, and say `<=1.0` or similar". 

In the case of AlertManager, it's currently API v2. Whether you want to tie the provider/consumer to API version, config file format, explicit application version (by querying `/status` on the AlertManager endpoint), or something completely different is up to you.

Reply:
1. Of course charm code is where you query for workload version. Here's an [example](https://github.com/canonical/alertmanager-operator/blob/main/src/charm.py#L110).

Now that we agree on (1) above, it seems even clearer than before that the argument to "version" is the actual version of the workload app.

afaiu, the interpretation "the provider can be a range" is inconsistent with the code snippet (and its intention) cited above (`SimpleSpec.match(Version.coerce(..))`).

Reply:
Yes, it’s on the Karma library. 

We never disagreed on that point. My point is that semver provides flexibility alongside constraints. This PR neither queries nor uses the API versioning. 

A docstring does not overrule an industry-wide understanding of what semver is and what its uses are. Grafana, for example, is using the API version, not the workload version. 

The point of the comment is that a hard coded string to the version number does not enforce semver compliance in any meaningful way nor provide feedback about major/minor versioning changes to consumers. The last part of my last comment provided multiple options. 

Implementation is up to you, but it is not as prescriptive as asserted by any means.

Reply:
A lot of this is defensive programming, and I mentioned the API version for a reason. Outside of the LMA use case, it does not cost a lot in code to say “what if?”

What if something uses the V1 API? How do we identify it? If the V2 API changes, how do we check?

If we don’t use API version, how do we know when the image updates or what is used to deploy if we don’t query?

Reply:
Using workload version is more defensive than API version, exactly for the "what ifs" you listed.

Reply:
Just to close the loop (we discussed on Mattermost anyway). A dynamic version getter is great.

Using this version does not acknowledge the use case of something like `etcd`, where the “what if I want to use API `v2` no matter what the image is?” is covered. 

For “defensive” programming, “what if the image used changes?” and “what if the API used changes?” is a choice. Especially in the case of “generic” provider/consumer libs where it should fulfill, say, MariaDB || MySQL, which will have different image versions but more or less identical APIs for common use cases (similar also for galera, percona, pgpool, etc). One is not “more” defensive than the other. 

I’m not trying to be nitpicky here, but `SemVer` supports both use cases, and since this is a public discussion early in the history of `ProviderBase`, I want to be crystal clear that both options are essentially equally valid depending on the use case and intended purpose of the library, since these may get indexed. 

I approved this a long time ago anyway, and these comments are becoming circular. Let’s merge it.